File: python/bwt_fm_index.py
"""
Burrows-Wheeler Transform (BWT) and FM-Index

The BWT is a reversible transformation that reorganizes a string to make it more compressible.
The FM-index uses the BWT to enable fast exact pattern matching and is foundational to
BWA, Bowtie, and HISAT2 aligners.
"""


def burrows_wheeler_transform(text):
    """
    Compute the Burrows-Wheeler Transform of a text.
    
    Args:
        text: Input string (must end with unique character like '$')
    
    Returns:
        str: BWT of the input text
    """
    # Add sentinel character if not present
    if not text.endswith('$'):
        text = text + '$'
    
    # Create all rotations
    rotations = [text[i:] + text[:i] for i in range(len(text))]
    
    # Sort rotations lexicographically
    rotations.sort()
    
    # BWT is the last column of sorted rotations
    bwt = ''.join(rotation[-1] for rotation in rotations)
    
    return bwt


def inverse_bwt(bwt):
    """
    Reverse the Burrows-Wheeler Transform.
    
    Args:
        bwt: BWT string
    
    Returns:
        str: Original text
    """
    n = len(bwt)
    
    # Create table with indices
    table = [(bwt[i], i) for i in range(n)]
    
    # Sort to get first column
    table.sort()
    
    # Follow the links to reconstruct
    result = []
    idx = 0  # Start with the row containing '$'
    
    for _ in range(n):
        result.append(table[idx][0])
        idx = table[idx][1]
    
    # Remove sentinel and return
    text = ''.join(result)
    return text[1:] + text[0]  # Rotate to put '$' at the end


class FMIndex:
    """
    FM-Index for efficient pattern matching using BWT.
    
    The FM-index enables backward search for exact pattern matching in O(m) time
    where m is the pattern length, independent of the text length.
    """
    
    def __init__(self, text):
        """
        Build FM-index from text.
        
        Args:
            text: Input text (string)
        """
        # Add sentinel if not present
        if not text.endswith('$'):
            text = text + '$'
        
        self.text = text
        self.bwt = burrows_wheeler_transform(text)
        self._build_index()
    
    def _build_index(self):
        """Build auxiliary data structures for FM-index."""
        n = len(self.bwt)
        
        # Count occurrences of each character
        self.counts = {}
        for char in self.bwt:
            self.counts[char] = self.counts.get(char, 0) + 1
        
        # Compute C array - number of characters lexicographically smaller
        self.C = {}
        sorted_chars = sorted(self.counts.keys())
        total = 0
        for char in sorted_chars:
            self.C[char] = total
            total += self.counts[char]
        
        # Build occurrence array (Occ)
        # Occ[char][i] = number of occurrences of char in bwt[0:i]
        self.occ = {char: [0] * (n + 1) for char in self.counts}
        
        for i, char in enumerate(self.bwt):
            for c in self.occ:
                self.occ[c][i + 1] = self.occ[c][i]
            self.occ[char][i + 1] += 1
    
    def count(self, pattern):
        """
        Count occurrences of pattern in the text.
        
        Args:
            pattern: Pattern to search for
        
        Returns:
            int: Number of occurrences
        """
        top, bottom = self._backward_search(pattern)
        if top > bottom:
            return 0
        return bottom - top + 1
    
    def _backward_search(self, pattern):
        """
        Perform backward search using FM-index.
        
        Args:
            pattern: Pattern to search for
        
        Returns:
            tuple: (top, bottom) range in suffix array
        """
        top = 0
        bottom = len(self.bwt) - 1
        
        # Process pattern from right to left
        for i in range(len(pattern) - 1, -1, -1):
            char = pattern[i]
            
            if char not in self.C:
                return (1, 0)  # Pattern not found
            
            # Update range using LF mapping
            top = self.C[char] + self.occ[char][top]
            bottom = self.C[char] + self.occ[char][bottom + 1] - 1
            
            if top > bottom:
                return (1, 0)  # Pattern not found
        
        return (top, bottom)
    
    def locate(self, pattern):
        """
        Find all positions where pattern occurs in text.
        
        Args:
            pattern: Pattern to search for
        
        Returns:
            list: List of positions where pattern occurs
        """
        top, bottom = self._backward_search(pattern)
        
        if top > bottom:
            return []
        
        # For simplicity, we reconstruct positions
        # In practice, this would use sampled suffix array
        positions = []
        
        # Create suffix array
        rotations = [(self.text[i:] + self.text[:i], i) for i in range(len(self.text))]
        rotations.sort()
        suffix_array = [rot[1] for rot in rotations]
        
        for i in range(top, bottom + 1):
            positions.append(suffix_array[i])
        
        return sorted(positions)
    
    def search(self, pattern):
        """
        Search for exact matches of pattern.
        
        Args:
            pattern: Pattern to search for
        
        Returns:
            dict: Search results with count and positions
        """
        count = self.count(pattern)
        positions = self.locate(pattern) if count > 0 else []
        
        return {
            'pattern': pattern,
            'count': count,
            'positions': positions
        }


if __name__ == "__main__":
    # Example usage
    text = "ACGTACGTACGT"
    
    print("Burrows-Wheeler Transform and FM-Index")
    print(f"Original text: {text}")
    
    # BWT
    bwt = burrows_wheeler_transform(text)
    print(f"BWT: {bwt}")
    
    # Inverse BWT
    original = inverse_bwt(bwt)
    print(f"Inverse BWT: {original}")
    print(f"Reconstruction correct: {original == text + '$'}")
    
    # FM-Index search
    print("\n--- FM-Index Search ---")
    fm_index = FMIndex(text)
    
    patterns = ["ACG", "CGT", "TAC", "XYZ"]
    for pattern in patterns:
        result = fm_index.search(pattern)
        print(f"\nPattern: {pattern}")
        print(f"Count: {result['count']}")
        if result['positions']:
            print(f"Positions: {result['positions']}")

File: cpp/bwt_fm_index.cpp
/**
 * Burrows-Wheeler Transform (BWT) and FM-Index
 * 
 * The BWT is a reversible transformation that reorganizes a string to make it more compressible.
 * The FM-index uses the BWT to enable fast exact pattern matching and is foundational to
 * BWA, Bowtie, and HISAT2 aligners.
 */

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <map>

using namespace std;

/**
 * Compute the Burrows-Wheeler Transform of a text.
 */
string burrows_wheeler_transform(string text) {
    // Add sentinel character if not present
    if (text.empty() || text.back() != '$') {
        text += '$';
    }
    
    int n = text.length();
    
    // Create all rotations
    vector<string> rotations;
    for (int i = 0; i < n; i++) {
        rotations.push_back(text.substr(i) + text.substr(0, i));
    }
    
    // Sort rotations lexicographically
    sort(rotations.begin(), rotations.end());
    
    // BWT is the last column of sorted rotations
    string bwt;
    for (const auto& rotation : rotations) {
        bwt += rotation.back();
    }
    
    return bwt;
}

/**
 * Reverse the Burrows-Wheeler Transform.
 */
string inverse_bwt(const string& bwt) {
    int n = bwt.length();
    
    // Create table with indices
    vector<pair<char, int>> table;
    for (int i = 0; i < n; i++) {
        table.push_back({bwt[i], i});
    }
    
    // Sort to get first column
    sort(table.begin(), table.end());
    
    // Follow the links to reconstruct
    string result;
    int idx = 0;  // Start with the row containing '$'
    
    for (int i = 0; i < n; i++) {
        result += table[idx].first;
        idx = table[idx].second;
    }
    
    // Rotate to put '$' at the end
    return result.substr(1) + result[0];
}

/**
 * FM-Index for efficient pattern matching using BWT.
 */
class FMIndex {
private:
    string text;
    string bwt;
    map<char, int> C;  // Number of characters lexicographically smaller
    map<char, vector<int>> occ;  // Occurrence array
    
    void build_index() {
        int n = bwt.length();
        
        // Count occurrences of each character
        map<char, int> counts;
        for (char c : bwt) {
            counts[c]++;
        }
        
        // Compute C array
        int total = 0;
        for (auto& p : counts) {
            C[p.first] = total;
            total += p.second;
        }
        
        // Build occurrence array (Occ)
        for (auto& p : counts) {
            occ[p.first] = vector<int>(n + 1, 0);
        }
        
        for (int i = 0; i < n; i++) {
            for (auto& p : occ) {
                p.second[i + 1] = p.second[i];
            }
            occ[bwt[i]][i + 1]++;
        }
    }
    
    pair<int, int> backward_search(const string& pattern) {
        int top = 0;
        int bottom = bwt.length() - 1;
        
        // Process pattern from right to left
        for (int i = pattern.length() - 1; i >= 0; i--) {
            char c = pattern[i];
            
            if (C.find(c) == C.end()) {
                return {1, 0};  // Pattern not found
            }
            
            // Update range using LF mapping
            top = C[c] + occ[c][top];
            bottom = C[c] + occ[c][bottom + 1] - 1;
            
            if (top > bottom) {
                return {1, 0};  // Pattern not found
            }
        }
        
        return {top, bottom};
    }

public:
    FMIndex(string input_text) {
        // Add sentinel if not present
        if (input_text.empty() || input_text.back() != '$') {
            input_text += '$';
        }
        
        text = input_text;
        bwt = burrows_wheeler_transform(input_text);
        build_index();
    }
    
    /**
     * Count occurrences of pattern in the text.
     */
    int count(const string& pattern) {
        auto [top, bottom] = backward_search(pattern);
        if (top > bottom) {
            return 0;
        }
        return bottom - top + 1;
    }
    
    /**
     * Find all positions where pattern occurs in text.
     */
    vector<int> locate(const string& pattern) {
        auto [top, bottom] = backward_search(pattern);
        
        if (top > bottom) {
            return {};
        }
        
        // Create suffix array for position lookup
        int n = text.length();
        vector<pair<string, int>> rotations;
        for (int i = 0; i < n; i++) {
            rotations.push_back({text.substr(i) + text.substr(0, i), i});
        }
        sort(rotations.begin(), rotations.end());
        
        vector<int> positions;
        for (int i = top; i <= bottom; i++) {
            positions.push_back(rotations[i].second);
        }
        
        sort(positions.begin(), positions.end());
        return positions;
    }
};

int main() {
    // Example usage
    string text = "ACGTACGTACGT";
    
    cout << "Burrows-Wheeler Transform and FM-Index" << endl;
    cout << "Original text: " << text << endl;
    
    // BWT
    string bwt = burrows_wheeler_transform(text);
    cout << "BWT: " << bwt << endl;
    
    // Inverse BWT
    string original = inverse_bwt(bwt);
    cout << "Inverse BWT: " << original << endl;
    cout << "Reconstruction correct: " << (original == text + "$" ? "true" : "false") << endl;
    
    // FM-Index search
    cout << "\n--- FM-Index Search ---" << endl;
    FMIndex fm_index(text);
    
    vector<string> patterns = {"ACG", "CGT", "TAC", "XYZ"};
    for (const auto& pattern : patterns) {
        int count = fm_index.count(pattern);
        cout << "\nPattern: " << pattern << endl;
        cout << "Count: " << count << endl;
        
        if (count > 0) {
            vector<int> positions = fm_index.locate(pattern);
            cout << "Positions: ";
            for (size_t i = 0; i < positions.size(); i++) {
                cout << positions[i];
                if (i < positions.size() - 1) cout << ", ";
            }
            cout << endl;
        }
    }
    
    return 0;
}
